### PR TITLE
ensure ucas imports are separated from reference data

### PR DESCRIPTION
**Pull request by the-daniel-rothig**
### Context

We want to do very regular UCAS data imports but want to keep our reference data (organisations, institutions, users, and mappings between those) intact

### Changes proposed in this pull request

Split the ucas import endpoint up by introducting a new reference data import endpoint.

### Guidance to review

This is mostly a refactor, splitting one method in DataController, IDataService and DataService into two. A corresponding ucas-importer PR is imminent.

Of particular note is a new integration test that checks if repeat imports succeed and leave the data as expected.

I held off renaming UcasInsitutions -> McInstitutions even though Mc* is essentially the reference data and Ucas* is the "throwaway" re-importable UCAS data for now, because our table and object naming could do with a slightly wider review (There is user data; course data; and org data... and there will be enrichment data. All four have different data lifecycles)

### CODE REVIEW

File: src/ManageCourses.Api/Model/UserPayload.cs [hunk 1]
```diff
@@ -0,0 +1,22 @@
+﻿using System.Collections.Generic;
+using GovUk.Education.ManageCourses.Domain.Models;
+
+namespace GovUk.Education.ManageCourses.Api.Model
+{
+    public class ReferenceDataPayload
```
Comment (MartinBentley) [inline src/ManageCourses.Api/Model/UserPayload.cs]: Can you rename the file to be the name of the class

File: src/ManageCourses.Api/Controllers/DataController.cs [hunk 2]
```diff
@@ -36,9 +36,22 @@ public OrganisationCourses ExportByOrganisation(string ucasCode)
...
+
+            //TODO return Ok/Fail in action result
+        }
+
+        /// <summary>
+        /// Imports the data.
+        /// </summary>
+        [HttpPost]
+        [Route("referencedata")]
+        public void ImportReferenceData([FromBody] ReferenceDataPayload payload)
```
Comment (defong) [inline src/ManageCourses.Api/Controllers/DataController.cs]: `

?`[Authorize(AuthenticationSchemes = BearerTokenApiKeyDefaults.AuthenticationScheme)]

File: src/ManageCourses.Api/Data/DataService.cs [hunk 3]
```diff
@@ -174,6 +157,30 @@ public void ProcessPayload(Payload payload)
                         Name = organisation.Name
                     }
                     );
+            }            
+            foreach (var institution in payload.Institutions)
+            {
+                _context.AddUcasInstitution(
+                    new UcasInstitution
```
Comment (defong) [inline src/ManageCourses.Api/Data/DataService.cs]: Maybe we should consider auto mapper

File: src/ManageCourses.ApiClient/Generated/ManageCoursesApiClient.cs [hunk 4]
```diff
@@ -193,18 +193,84 @@ public System.Threading.Tasks.Task<OrganisationCourses> Data_ExportByOrganisatio
...
+        public System.Threading.Tasks.Task Data_ImportAsync(UcasPayload payload)
         {
             return Data_ImportAsync(payload, System.Threading.CancellationToken.None);
         }
     
         /// <summary>Imports the data.</summary>
         /// <exception cref="SwaggerException">A server side error occurred.</exception>
         /// <param name="cancellationToken">A cancellation token that can be used by other objects or threads to receive notice of cancellation.</param>
-        public async System.Threading.Tasks.Task Data_ImportAsync(Payload payload, System.Threading.CancellationToken cancellationToken)
+        public async System.Threading.Tasks.Task Data_ImportAsync(UcasPayload payload, System.Threading.CancellationToken cancellationToken)
```
Comment (defong) [inline src/ManageCourses.ApiClient/Generated/ManageCoursesApiClient.cs]: Do we actually need this to be commited?

File: tests/ManageCourses.Tests/SmokeTests/SystemUnderTest.cs [hunk 5]
```diff
@@ -160,8 +160,9 @@ public Task<string> GetAccessTokenAsync()
             }
         }
         private static class TestData {
```
Comment (defong) [inline tests/ManageCourses.Tests/SmokeTests/SystemUnderTest.cs]: Can we move this class out?

Comment (the-daniel-rothig) [reply tests/ManageCourses.Tests/SmokeTests/SystemUnderTest.cs]: thx done

File: src/ManageCourses.ApiClient/Generated/ManageCoursesApiClient.cs (hunk 4 above)
Comment (the-daniel-rothig) [reply src/ManageCourses.ApiClient/Generated/ManageCoursesApiClient.cs]: nope thx - fixed

File: src/ManageCourses.Api/Data/DataService.cs (hunk 3 above)
Comment (the-daniel-rothig) [reply src/ManageCourses.Api/Data/DataService.cs]: Possibly, but I don't want to do this in this PR

File: src/ManageCourses.Api/Controllers/DataController.cs (hunk 2 above)
Comment (the-daniel-rothig) [reply src/ManageCourses.Api/Controllers/DataController.cs]: oops thx

File: tests/ManageCourses.Tests/Integration/DataServiceTests.cs [hunk 6]
```diff
@@ -31,27 +31,42 @@ public class DataServiceTests : ManageCoursesDbContextIntegrationBase
...
-        {
-            var payload = GetPayload();
-
-            Subject.ProcessPayload(payload);
+        {            
+            var userPayload = GetUserPayload();
+            Subject.ProcessReferencePayload(userPayload);
+            
+            var payload = GetUcasPayload();
+            Subject.ProcessUcasPayload(payload);
```
[the remark this answers is not in the document]
Comment (the-daniel-rothig) [reply tests/ManageCourses.Tests/Integration/DataServiceTests.cs]: yep :)

File: src/ManageCourses.Api/Model/UserPayload.cs (hunk 1 above)
Comment (the-daniel-rothig) [reply src/ManageCourses.Api/Model/UserPayload.cs]: oops thx

**Comment by defong** [comment]
SmokeTests/TestData.cs(5,23): error CS0246: The type or namespace name 'ReferenceDataPayload' could not be found (are you missing a using directive or an assembly reference?) [/home/travis/build/DFE-Digital/manage-courses-api/tests/ManageCourses.Tests/ManageCourses.Tests.csproj]
SmokeTests/TestData.cs(35,23): error CS0246: The type or namespace name 'UcasPayload' could not be found (are you missing a using directive or an assembly reference?) [/home/travis/build/DFE-Digital/manage-courses-api/tests/ManageCourses.Tests/ManageCourses.Tests.csproj]
SmokeTests/TestData.cs(52,24): error CS0246: The type or namespace name 'ObservableCollection<>' could not be found (are you missing a using directive or an assembly reference?) [/home/travis/build/DFE-Digital/manage-courses-api/tests/ManageCourses.Tests/ManageCourses.Tests.csproj]

**Pull request merged**